### PR TITLE
Add ability to customize ping messages with custom data, see iss…

### LINKED ISSUE #941: How to send customized ping message on connectionLostTimeout interval
**Describe what you would like to know or do**
How to send a custom Ping message on `connectionLostTimeout`?

**Describe the solution you'd considered**
For creating a WS client, I am extending `WebSocketClient`. I did `setConnectionLostTimeout` to 30 seconds. and override the `sendPing()` method https://github.com/TooTallNate/Java-WebSocket/blob/a2c63db70cf33ff838acdfdedf34f50eb680eae7/src/main/java/org/java_websocket/client/WebSocketClient.java#L439 

```
@Override
  public void sendPing() {
    log.info("Sending ping");
    JsonObject pingObj = new JsonObject();
    pingObj.addProperty("id", "NGenerated guid id");
    pingObj.addProperty("type", "ping");
    this.send(pingObj.toString());
  }
```

I thought it would be simple, just to override `sendPing()`. But it is not the case. My overridden method never gets invoked. Rather it calls engine's `sendPing()` https://github.com/TooTallNate/Java-WebSocket/blob/a2c63db70cf33ff838acdfdedf34f50eb680eae7/src/main/java/org/java_websocket/WebSocketImpl.java#L661

**Additional context**
It's not causing any connection close issue, but I am just looking for a way o customize the send ping message.

### PR DESCRIPTION
**Pull request by haruntuncay**
## Description
Added a method called `onPreparePing` which provides a way for users customize their ping messages with custom data that they want to send.

## Related Issue

Fixes #941 

## Motivation and Context
Solves the problem where users were not able to customize ping messages. In this context, customizing means filling ping messages with custom payload data.

## How Has This Been Tested?
Added a test file.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by marci4** [comment]
@PhilipRoman thx for the review and @haruntuncay thx for the contribution

File: src/main/java/org/java_websocket/WebSocketListener.java [hunk 1]
```diff
@@ -169,6 +170,13 @@
 	 */
 	void onWebsocketPing( WebSocket conn, Framedata f );
 
+	/**
+	 * Called just before a ping frame is sent, in order to allow users to customize their ping frame data.
+	 *
+	 * @return PingFrame to be sent.
+	 */
+	PingFrame onPreparePing(WebSocket conn );
```
Comment (marci4) [inline src/main/java/org/java_websocket/WebSocketListener.java]: Could you please include the argument?

File: src/main/java/org/java_websocket/WebSocketAdapter.java [hunk 2]
```diff
@@ -85,4 +87,17 @@ public void onWebsocketPing( WebSocket conn, Framedata f ) {
...
 	}
+
+	/**
+	 * Default implementation for onPreparePing, returns a (cached) PingFrame that has no application data.
+	 *
+	 * @see org.java_websocket.WebSocketListener#onPreparePing(WebSocket)
+	 * @return PingFrame to be sent.
+	 */
+	@Override
+	public PingFrame onPreparePing(WebSocket conn) {
```
Comment (marci4) [inline src/main/java/org/java_websocket/WebSocketAdapter.java]: Could you please include the argument?

File: src/main/java/org/java_websocket/WebSocketAdapter.java [hunk 3]
```diff
@@ -40,6 +40,8 @@
  **/
 public abstract class WebSocketAdapter implements WebSocketListener {
 
+	private PingFrame pingFrame;
```
Comment (marci4) [inline src/main/java/org/java_websocket/WebSocketAdapter.java]: Do you even need to cache it anymore? Should the user not be able to send each cycle specific data?

Comment (PhilipRoman) [reply src/main/java/org/java_websocket/WebSocketAdapter.java]: I believe that the cache is used only for the default implementation. If the user overrides this method, the frame will not be cached.

Comment (haruntuncay) [reply src/main/java/org/java_websocket/WebSocketAdapter.java]: I think most people will not be overriding the default implementation, so a cache could be useful. Atleast that's what I thought, though I can change it if you like.

Comment (marci4) [reply src/main/java/org/java_websocket/WebSocketAdapter.java]: Yes, missed it.

**Comment by haruntuncay** [comment]
@PhilipRoman , @marci4 , Thank you both for the review and merge.

**Pull request merged**